# Módulo 5:
### Módulos, paquetes, cadenas, métodos de listas y excepciones.

En este módulo, aprenderás acerca de:

- Módulos de Python: su lógica, su función, cómo importarlos de diferentes maneras y presentar el contenido de algunos módulos estándar proporcionados por Python.
- La forma en que los módulos se unen para formar paquetes.
- El concepto de una excepción y su implementación en Python, incluida la instrucción `try`-`except`, con sus aplicaciones y la instrucción `raise`.
- Cadenas y sus métodos específicos, junto con sus similitudes y diferencias en comparación con las listas.

## 4.1 Módulos

Si se desea que un proyecto de software se complete con éxito, se deben tener los medios que le permitan:

- Se deben dividir todas las tareas entre los desarrolladores.
- Después, unir todas las partes creadas en un todo funcional.

Por ejemplo, un determinado proyecto se puede dividir en dos partes principales:

- La interfaz de usuario (la parte que se comunica con el usuario mediante widgets y una pantalla gráfica).
- La lógica (la parte que procesa los datos y produce resultados).

Cada una de estas partes se puede (muy probablemente) dividir en otras más pequeñas, y así sucesivamente. Tal proceso a menudo se denomina **descomposición**. 

El manejo de los módulos consta de dos cuestiones diferentes:

- El primero (probablemente el más común) ocurre cuando se desea utilizar un módulo ya existente, escrito por otra persona o creado por el programador mismo en algún proyecto complejo: en este caso, se considera al programador como el **usuario** del módulo.
- El segundo ocurre cuando se desea crear un nuevo módulo, ya sea para uso propio o para facilitar la vida de otros programadores: aquí eres el **proveedor** del módulo.

En primer lugar, un módulo se identifica por su **nombre**. Si se desea utilizar cualquier módulo, se necesita saber su nombre. Se entrega una cantidad (bastante grande) de módulos junto con Python. Se puede pensar en ellos como una especie de "equipo extra de Python".

Todos estos módulos, junto con las funciones integradas, forman la **Biblioteca estándar de Python** - un tipo especial de biblioteca donde los módulos desempeñan el papel de libros (incluso podemos decir que las carpetas desempeñan el papel de estanterías). Si deseas ver la lista completa de todos los "volúmenes" recopilados en esa biblioteca, se puede encontrar aquí: https://docs.python.org/3/library/index.html.

Cada módulo consta de entidades (como un libro consta de capítulos). Estas entidades pueden ser funciones, variables, constantes, clases y objetos. Si se sabe cómo acceder a un módulo en particular, se puede utilizar cualquiera de las entidades que almacena.

### 4.1.1 Importando un módulo

#### Primer método

Para que un módulo sea utilizable, hay que importarlo (piensa en ello como sacar un libro del estante). La importación de un módulo se realiza mediante una instrucción llamada `import`. N.B.: `import` es también una palabra reservada (con todas sus implicaciones).

In [ ]:
import math # importando el módulo math
import math, sys ## importando los módulos math y sys

Un **namespace** es un espacio (entendido en un contexto no físico) en el que existen algunos nombres y los nombres no entran en conflicto entre sí (es decir, no hay dos objetos diferentes con el mismo nombre). Podemos decir que cada grupo social es un namespace - el grupo tiende a nombrar a cada uno de sus miembros de una manera única (por ejemplo, los padres no darán a sus hijos los mismos nombres).

**Dentro de un determinado namespace, cada nombre debe permanecer único**. Esto puede significar que algunos nombres pueden desaparecer cuando cualquier otra entidad de un nombre ya conocido ingresa al namespace. Mostraremos cómo funciona y cómo controlarlo, pero primero, volvamos a las importaciones.

Si el módulo de un nombre especificado **existe y es accesible** (**un módulo es de hecho un archivo fuente de Python**), **Python importa su contenido, se hacen conocidos todos los nombres definidos en el módulo, pero no ingresan al namespace del código**.

In [ ]:
print('math.pi:',math.pi)
print('math.sin(0):',math.sin(0))
print('math.sin(math.pi):',math.sin(math.pi))

math.pi: 3.141592653589793
math.sin(0): 0.0
math.sin(math.pi): 1.2246467991473532e-16


In [ ]:
import math
pi = 3.14 
print('pi\t:\t', pi) # ¡atención aquí!
print('math.pi\t:\t', math.pi)

pi	:	 3.14
math.pi	:	 3.141592653589793


#### Segundo método

En el segundo método, la sintaxis del `import` señala con precisión qué entidad (o entidades) del módulo son aceptables en el código:

In [ ]:
from math import pi
print('pi:',pi)

from math import sin, pi
print('sin(pi/2):',sin(pi/2))

pi: 3.141592653589793
sin(pi/2): 1.0


In [ ]:
# Restart runtime
from math import pi # solo importa pi de math 
pi = 3.14 
print('pi\t:\t', pi) 
print('math.pi\t:\t', math.pi) # ¡atención aquí!

pi	:	 3.14


NameError: ignored

#### Tercer método

En el tercer método, la sintaxis del `import` es una forma más agresiva que la presentada anteriormente: `from module import *`

Como puedes ver, el nombre de una entidad (o la lista de nombres de entidades) se reemplaza con un sólo asterisco (*). Tal instrucción importa todas las entidades del módulo indicado.

- ¿Es conveniente? Sí, lo es, ya que libera del deber de enumerar todos los nombres que se necesiten.
- ¿Es inseguro? Sí, a menos que conozca todos los nombres proporcionados por el módulo, es posible que no puedas evitar conflictos de nombres. Trata esto como una solución temporal e intenta no usarlo en un código regular.


#### Cuarto método

Si se importa un módulo y no se está conforme con el nombre del módulo en particular (por ejemplo, sí es el mismo que el de una de sus entidades ya definidas) puede dársele otro nombre: esto se llama **aliasing** o **renombrado**. Aliasing (renombrado) **hace que el módulo se identifique con un nombre diferente al original**.

La creación de un alias se realiza junto con la importación del módulo, y exige la siguiente forma de la instrucción `import`.

In [ ]:
import module as alias

El "module" identifica el nombre del módulo original mientras que el "alias" es el nombre que se desea usar en lugar del original. N.B.: `as` es una palabra reservada; después de la ejecución exitosa de una importación con alias, el nombre original del módulo se vuelve inaccesible y no debe ser utilizado.

A su vez, cuando usa la variante `from module import name` y se necesita cambiar el nombre de la entidad, se crea un alias para la entidad. Esto hará que el nombre sea reemplazado por el alias que se elija. Así es como se puede hacer:

In [ ]:
from module import nombre as alias

Como anteriormente, el nombre original (sin alias) se vuelve inaccesible. La frase `nombre as alias` puede repetirse: emplea comas para separar las frases, como a continuación:

In [ ]:
from math import pi as PI, sin as seno
print('seno(PI/2):',seno(PI/2))

sine(PI/2): 1.0


### 4.1.2 Módulos estándar

Antes de comenzar a revisar algunos módulos estándar de Python, veamos la función `dir()`. No tiene nada que ver con el comando `dir` de las terminales de Windows o Unix. El comando `dir()` no muestra el contenido de un directorio o carpeta de disco, pero no se puede negar que hace algo similar: puede revelar todos los nombres proporcionados a través de un módulo en particular.

Hay una condición: el módulo debe haberse importado previamente como un todo (es decir, utilizar la instrucción `import module` - `from module` no es suficiente).

La función devuelve una **lista ordenada alfabéticamente** la cual contiene todos los nombres de las entidades disponibles en el módulo: `dir(module)`

N.B: Si el nombre del módulo tiene un alias, debe usar el alias, no el nombre original. 

Usar la función dentro de un script normal no tiene mucho sentido, pero aún así, es posible. Por ejemplo, se puede ejecutar el siguiente código para imprimir los nombres de todas las entidades dentro del módulo `math`:

In [ ]:
import math

for name in dir(math):
    print(name, end="\t")

__doc__	__loader__	__name__	__package__	__spec__	acos	acosh	asin	asinh	atan	atan2	atanh	ceil	copysign	cos	cosh	degrees	e	erf	erfc	exp	expm1	fabs	factorial	floor	fmod	frexp	fsum	gamma	gcd	hypot	inf	isclose	isfinite	isinf	isnan	ldexp	lgamma	log	log10	log1p	log2	modf	nan	pi	pow	radians	sin	sinh	sqrt	tan	tanh	tau	trunc	

¿Has notado los nombres extraños que comienzan con `__` al inicio de la lista? Se hablará más sobre ellos cuando hablemos sobre los problemas relacionados con la escritura de módulos propios.

Algunos de los nombres pueden traer recuerdos de las lecciones de matemáticas, y probablemente no tendrás ningún problema en adivinar su significado.

El emplear la función `dir()` dentro de un código puede no parecer muy útil; por lo general, se desea conocer el contenido de un módulo en particular antes de escribir y ejecutar el código.

Afortunadamente, se puede ejecutar la función directamente en la consola de Python (IDLE), sin necesidad de escribir y ejecutar un script por separado.

Así es como se puede hacer:

In [ ]:
import math
dir(math)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'pi',
 'pow',
 'radians',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc']

#### El módulo `math`

Comencemos con una vista previa de algunas de las funciones proporcionadas por el módulo `math`. El primer grupo de ellas están relacionadas con trigonometría:

- `sin(x)` → el seno de x.
- `cos(x)` → el coseno de x.
- `tan(x)` → la tangente de x.

Todas estas funciones toman un argumento (una medida de ángulo expresada en radianes) y devuelven el resultado apropiado (ten cuidado con `tan()` - no todos los argumentos son aceptados).

También están sus versiones inversas:

- `asin(x)` → el arcoseno de x.
- `acos(x)` → el arcocoseno de x.
- `atan(x)` → el arcotangente de x.

Estas funciones toman un argumento (verifica que sea correcto) y devuelven una medida de un ángulo en radianes.

Para trabajar eficazmente con mediciones de ángulos, el módulo `math` proporciona las siguientes entidades:

- `pi` → una constante con un valor que es una aproximación de π.
- `radians(x)` → una función que convierte x de grados a radianes.
- `degrees(x)` → actuando en el otro sentido (de radianes a grados).

Además de las funciones circulares (enumeradas anteriormente), el módulo `math` también contiene un conjunto de sus análogos hiperbólicos:

- `sinh(x)` → el seno hiperbólico.
- `cosh(x)` → el coseno hiperbólico.
- `tanh(x)` → la tangente hiperbólico.
- `asinh(x)` → el arcoseno hiperbólico.
- `acosh(x)` → el arcocoseno hiperbólico.
- `atanh(x)` → el arcotangente hiperbólico.

Existe otro grupo de las funciones `math` relacionadas con la exponenciación:

- `e` → una constante con un valor que es una aproximación del número de Euler (e).
- `exp(x)` → encontrar el valor de $e^x$.
- `log(x)` → el logaritmo natural de x.
- `log(x, b)` → el logaritmo de x con base b.
- `log10(x)` → el logaritmo decimal de x (más preciso que `log(x, 10)`).
- `log2(x)` → el logaritmo binario de x (más preciso que `log(x, 2)`).

N.B: la función `pow()`:

- `pow(x, y)` → encontrar el valor de $x^y$ (toma en cuenta los dominios).
Esta es una función incorporada y no se tiene que importar.

El último grupo consta de algunas funciones de propósito general como:

- `ceil(x)` → devuelve el entero más pequeño mayor o igual que x.
- `floor(x)` → el entero más grande menor o igual que x.
- `trunc(x)` → el valor de x truncado a un entero (ten cuidado, no es equivalente a ceil o floor).
- `factorial(x)` → devuelve $x!$ (x tiene que ser un valor entero y no negativo.
- `hypot(x, y)` → devuelve la longitud de la hipotenusa de un triángulo rectángulo con las longitudes de los catetos iguales a x e y (lo mismo que `sqrt(pow(x, 2) + pow(y, 2)`) pero más preciso).

#### El módulo `random`

Otro módulo que vale la pena mencionar es el que se llama `random`. El cual ofrece algunos mecanismos que permiten operar con **números pseudoaleatorios**.

Toma en cuenta el prefijo **pseudo** - los números generados por los módulos pueden parecer aleatorios en el sentido de que no se pueden predecir, pero no hay que olvidar que todos se calculan utilizando algoritmos muy refinados.

Los algoritmos no son aleatorios, son deterministas y predecibles. Solo aquellos procesos físicos que se salgan completamente de nuestro control (como la intensidad de la radiación cósmica) pueden usarse como fuente de datos aleatorios reales. Los datos producidos por computadoras deterministas no pueden ser aleatorios de ninguna manera.

Un generador de números aleatorios toma un valor llamado **semilla**, lo trata como un valor de entrada, calcula un número "aleatorio" basado en él (el método depende de un algoritmo elegido) y produce una **nueva semilla**.

La duración de un ciclo en el que todos los valores semilla son únicos puede ser muy largo, pero no es infinito: tarde o temprano los valores iniciales comenzarán a repetirse y los valores generadores también se repetirán. Esto es normal. Es una característica, no un error.

El valor de la semilla inicial, establecido durante el inicio del programa, determina el orden en que aparecerán los valores generados.

El factor aleatorio del proceso puede ser aumentado al establecer la semilla tomando un número de la hora actual - esto puede garantizar que cada lanzamiento del programa comience desde un valor semilla diferente (por lo tanto, usará diferentes números aleatorios). Afortunadamente, Python realiza dicha inicialización al importar el módulo.

La función general llamada `random()` (no debe confundirse con el nombre del módulo) produce un número flotante `x` entre el rango `(0.0, 1.0)` - en otras palabras: (0.0 <= x < 1.0).

El sgte. programa producirá cinco valores pseudoaleatorios, ya que sus valores están determinados por el valor semilla (un valor impredecible) actual, no se pueden adivinar.

In [ ]:
from random import random

for i in range(5):
    print(random())

0.32015836040826195
0.2846077334072499
0.7624767762048694
0.24494027249722494
0.9478414048917418


La función `seed()` es capaz de establecer la semilla del generador. Te mostraremos dos de sus variantes:

- `seed()` - establece la semilla con la hora actual.
- `seed(int_value)` - establece la semilla con el valor entero int_value.

Hemos modificado el programa anterior; de hecho, hemos eliminado cualquier rastro de aleatoriedad del código:

In [ ]:
from random import random, seed

seed(0)

for i in range(5):
    print(random())

0.8444218515250481
0.7579544029403025
0.420571580830845
0.25891675029296335
0.5112747213686085


Debido al hecho de que la semilla siempre se establece con el mismo valor, la secuencia de valores generados siempre se ve igual. Ejecuta el programa. Esto es lo que tenemos:

`0.844421851525
0.75795440294
0.420571580831
0.258916750293
0.511274721369`

N.B.: sus valores pueden ser ligeramente diferentes si tu sistema utiliza aritmética de punto flotante más precisa o menos precisa, pero la diferencia se verá bastante lejos del punto decimal.

Si deseas valores aleatorios enteros, una de las siguientes funciones encajaría mejor:

- `randrange(fin)`
- `randrange(inico, fin)`
- `randrange(inicio, fin, incremento)`
- `randint(izquierda, derecha)`

Las primeras tres invocaciones generarán un número entero tomado (pseudoaleatoriamente) del rango:

- `range(fin)`
- `range(inicio, fin)`
- `range(inicio, fin, incremento)`

Toma en cuenta la **exclusión implícita del lado derecho**.

La última función es equivalente a `randrange(izquierda, derecha+1)` - genera el valor entero `i`, el cual cae en el rango [izquierda, derecha] (sin exclusión en el lado derecho).

Observa el código en el editor. Este programa generará una línea que consta de tres ceros y un cero o un uno en el cuarto lugar.

In [ ]:
from random import randrange, randint

print(randrange(1), end=' ')
print(randrange(0, 1), end=' ')
print(randrange(0, 1, 1), end=' ')
print(randint(0, 1))

0 0 0 1


Las funciones anteriores tienen una desventaja importante: pueden producir valores repetidos incluso si el número de invocaciones posteriores no es mayor que el rango especificado.

Observa el código en el editor. Es muy probable que el programa genere un conjunto de números en el que algunos elementos no sean únicos.

In [ ]:
for i in range(10):
    print(randint(1, 10), end=',')

10,4,9,3,5,3,2,10,5,9,

Como puedes ver, esta no es una buena herramienta para generar números para la lotería. Afortunadamente, existe una mejor solución que escribir tu propio código para verificar la singularidad de los números "sorteados". Es una función con el nombre de `choice`:

- `choice(secuencia)`
- `sample(secuencia, elementos_a_elegir=1)`

La primera variante elige un elemento "aleatorio" de la secuencia de entrada y lo devuelve. El segundo crea una lista (una muestra) que consta del elemento `elementos_a_elegir` (que por defecto es 1) "sorteado" de la secuencia de entrada.

En otras palabras, la función elige algunos de los elementos de entrada, devolviendo una lista con la elección. Los elementos de la muestra se colocan en orden aleatorio. Nota que `elementos_a_elegir` no debe ser mayor que la longitud de la secuencia de entrada.

Observa el código a continuación:

In [ ]:
from random import choice, sample

lst = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

print(choice(lst))
print(sample(lst, 5))
print(sample(lst, 10))

10
[3, 5, 2, 6, 1]
[6, 8, 2, 3, 4, 7, 9, 5, 10, 1]


Nuevamente, la salida del programa no es predecible.

#### El módulo `platform`

A veces, puede ser necesario encontrar información no relacionada con Python. Por ejemplo, es posible que necesites conocer la ubicación de tu programa dentro del entorno de la computadora.

Imagina el entorno de tu programa como una pirámide que consta de varias capas o plataformas.

Las capas son:

- El código (en ejecución) se encuentra en la parte superior.
- Python (mejor dicho, su entorno de ejecución) se encuentra directamente debajo de él.
- La siguiente capa de la pirámide se llena con el SO (sistema operativo): el entorno de Python proporciona algunas de sus funcionalidades utilizando los servicios del sistema operativo. Python, aunque es muy potente, no es omnipotente: se ve obligado a usar muchos ayudantes si va a procesar archivos o comunicarse con dispositivos físicos.
- La capa más inferior es el hardware: el procesador (o procesadores), las interfaces de red, los dispositivos de interfaz humana (ratones, teclados, etc.) y toda otra maquinaria necesaria para hacer funcionar la computadora: el sistema operativo sabe cómo emplearlos y utiliza muchos trucos para trabajar con todas las partes en un ritmo constante.

Esto significa que algunas de las acciones del programa tienen que recorrer un largo camino para ejecutarse con éxito, imagina que:

- **Tu código** quiere crear un archivo, por lo que invoca una de las funciones de Python.
- **Python** acepta la orden, la reorganiza para cumplir con los requisitos del sistema operativo local (es como poner el sello "aprobado" en una solicitud) y lo envía.
- El **SO** comprueba si la solicitud es razonable y válida (por ejemplo, si el nombre del archivo se ajusta a algunas reglas de sintaxis) e intenta crear el archivo. Tal operación, aparentemente es muy simple, no es atómica: consiste de muchos pasos menores tomados por...
- El **hardware**, el cual es responsable de activar los dispositivos de almacenamiento (disco duro, dispositivos de estado sólido, etc.) para satisfacer las necesidades del sistema operativo.

Por lo general, no eres consciente de todo ese alboroto: quieres que se cree el archivo y eso es todo.

Pero a veces quieres saber más, por ejemplo, el nombre del sistema operativo que aloja Python y algunas características que describen el hardware que aloja el sistema operativo.

Hay un módulo que proporciona algunos medios para permitir saber dónde se encuentra y qué componentes funcionan. El módulo se llama `platform`. Veamos algunas de las funciones que brinda.

El módulo `platform` permite acceder a los datos de la plataforma subyacente, es decir, hardware, sistema operativo e información sobre la versión del intérprete.

Existe también una función que puede mostrar todas las capas subyacentes en un solo vistazo, llamada `platform`. Simplemente devuelve una cadena que describe el entorno; por lo tanto, su salida está más dirigida a los humanos que al procesamiento automatizado (lo veremos pronto).

Así es como se puede invocar: `platform(aliased = False, terse = False)`.

Ahora:

- `aliased` → cuando se establece a True (o cualquier valor distinto de cero) puede hacer que la función presente los nombres de capa subyacentes alternativos en lugar de los comunes.
- `terse` → cuando se establece a True (o cualquier valor distinto de cero) puede convencer a la función de presentar una forma más breve del resultado (si lo fuera posible).

In [ ]:
from platform import platform

print(platform())
print(platform(1))
print(platform(0, 1))

Linux-4.19.104+-x86_64-with-Ubuntu-18.04-bionic
Linux-4.19.104+-x86_64-with-Ubuntu-18.04-bionic
Linux-4.19.104+-x86_64-with-glibc2.25


A veces, es posible que solo se desee conocer el nombre genérico del procesador que ejecuta el sistema operativo junto con Python y el código, una función llamada `machine()` te lo dirá. Como anteriormente, la función devuelve una cadena.

In [ ]:
from platform import machine

print(machine())

x86_64


La función `processor()` devuelve una cadena con el nombre real del procesador (si lo fuese posible).

In [ ]:
from platform import processor

print(processor())

x86_64


Una función llamada `system()` devuelve el nombre genérico del sistema operativo en una cadena.

In [ ]:
from platform import system

print(system())

Linux


La versión del sistema operativo se proporciona como una cadena por la función `version()`.

In [ ]:
from platform import version

print(version())

#1 SMP Wed Feb 19 05:26:34 PST 2020


Si necesitas saber qué versión de Python está ejecutando tu código, puedes verificarlo utilizando una serie de funciones dedicadas, aquí hay dos de ellas:

- `python_implementation()` → devuelve una cadena que denota la implementación de Python (espera CPython aquí, a menos que decidas utilizar cualquier rama de Python no canónica).
- `python_version_tuple()` → devuelve una tupla de tres elementos la cual contiene:
 - la parte **mayor** de la versión de Python.
 - la parte **menor**,
 - el número de nivel del **patch**.

In [ ]:
from platform import python_implementation, python_version_tuple

print(python_implementation())

for atr in python_version_tuple():
    print(atr)

CPython
3
6
9


#### Índice del módulo de Python

Aquí solo hemos cubierto los conceptos básicos de los módulos de Python. Los módulos de Python conforman su propio universo, en el que Python es solo una galaxia, y nos aventuraríamos a decir que explorar las profundidades de estos módulos puede llevar mucho más tiempo que familiarizarse con Python "puro".

Además, la comunidad de Python en todo el mundo crea y mantiene cientos de módulos adicionales utilizados en aplicaciones muy específicas como la genética, la psicología o incluso la astrología.

Estos módulos no están (y no serán) distribuidos junto con Python, o a través de canales oficiales, lo que hace que el universo de Python sea más amplio, casi infinito.

Puedes leer sobre todos los módulos estándar de Python aquí: https://docs.python.org/3/py-modindex.html.

No te preocupes, no necesitarás todos estos módulos. Muchos de ellos son muy específicos. Todo lo que se necesita hacer es encontrar los módulos que se desean y aprender a cómo usarlos.

### 4.1.3 Creando módulos y paquetes propios

Escribir tus propios módulos no difiere mucho de escribir scripts comunes. Existen algunos aspectos específicos que se deben tomar en cuenta, pero definitivamente no es algo complicado. Lo verás pronto.

Resumamos algunos aspectos importantes:

- Un **módulo** es un **contenedor lleno de funciones** - puedes empaquetar tantas funciones como desees en un módulo y distribuirlo por todo el mundo.
- Por supuesto, no es una buena idea mezclar funciones con diferentes áreas de aplicación dentro de un módulo (al igual que en una biblioteca: nadie espera que los trabajos científicos se incluyan entre los cómics), así que se deben agrupar las funciones cuidadosamente y asignar un nombre claro e intuitivo al módulo que las contiene (por ejemplo, no le des el nombre videojuegos a un módulo que contiene funciones destinadas a particionar y formatear discos duros).
- Crear muchos módulos puede causar desorden: tarde que temprano querrás agrupar tus módulos de la misma manera que previamente has agrupado funciones: ¿Existe un contenedor más general que un módulo?.
- Sí lo hay, es un **paquete**: en el mundo de los módulos, un paquete juega un papel similar al de una carpeta o directorio en el mundo de los archivos.

En esta sección, trabajarás localmente en tu máquina. Comencemos desde cero, de la siguiente manera:

Se necesitan dos archivos para realizar estos experimentos. Uno de ellos será el módulo en sí. Está vacío ahora. No te preocupes, lo vas a llenar con el código real.

El archivo lleva por nombre `module.py.` No muy creativo, pero es simple y claro.

El segundo archivo contiene el código usando el nuevo módulo. Su nombre es `main.py`. Su contenido es muy breve hasta ahora: `import module`

N.B.: ambos archivos deben estar ubicados en la misma carpeta. Te recomendamos crear una carpeta nueva y vacía para ambos archivos. Esto hará que las cosas sean más fáciles.

Inicia el IDLE y ejecuta el archivo `main.py`. ¿Qué ves?

No deberías ver nada. Esto significa que Python ha importado con éxito el contenido del archivo `module.py`. No importa que el módulo esté vacío por ahora. El primer paso ya está hecho, pero antes de dar el siguiente paso, queremos que eches un vistazo a la carpeta en la que se encuentran ambos archivos.

¿Notas algo interesante?

Ha aparecido una nueva subcarpeta, ¿puedes verla? Su nombre es `__pycache__`. Echa un vistazo adentro. ¿Que ves?

Hay un archivo llamado (más o menos) `module.cpython-xy.pyc` donde x e y son dígitos derivados de tu versión de Python (por ejemplo, serán 3 y 4 si utilizas Python 3.4).

El nombre del archivo es el mismo que el de tu módulo. La parte posterior al primer punto dice qué implementación de Python ha creado el archivo (CPython) y su número de versión. La ultima parte (pyc) viene de las palabras Python y compilado.

Puedes mirar dentro del archivo: el contenido es completamente ilegible para los humanos. Tiene que ser así, ya que el archivo está destinado solo para uso de Python.

Cuando Python importa un módulo por primera vez, **traduce el contenido a una forma "semi" compilada**. El archivo no contiene código en lenguaje máquina: **es código semi-compilado** interno de Python, listo para ser ejecutado por el intérprete de Python. Como tal archivo no requiere tantas comprobaciones como las de un archivo fuente, la ejecución comienza más rápido y también se ejecuta más rápido.

Gracias a eso, cada importación posterior será más rápida que interpretar el código fuente desde cero.

Python puede verificar si el archivo fuente del módulo ha sido modificado (en este caso, el archivo pyc será reconstruido) o no (cuando el archivo pyc pueda ser ejecutado al instante). Este proceso es completamente automático y transparente, no se tiene que estar tomando en cuenta.

Ahora hemos puesto algo en el archivo del `module.py`: `print('I like to be a module.')`

¿Puedes notar alguna diferencia entre un módulo y un script ordinario? No hay ninguna hasta ahora.

Es posible ejecutar este archivo como cualquier otro script. Pruébalo por ti mismo. ¿Qué es lo que pasa? Deberías de ver la siguiente línea dentro de tu consola: `I like to be a module.`

Volvamos al archivo `main.py`. Ejecuta el archivo. ¿Que ves? Con suerte, verás algo como esto: `I like to be a module.`

¿Qué significa realmente?

**Cuando un módulo es importado, su contenido es ejecutado implícitamente por Python**. Le da al módulo la oportunidad de inicializar algunos de sus aspectos internos (por ejemplo, puede asignar a algunas variables valores útiles). Nota: la **inicialización se realiza solo una vez**, cuando se produce la primera importación, por lo que las asignaciones realizadas por el módulo no se repiten innecesariamente.

Imagina el siguiente contexto:

- Existe un módulo llamado mod1.
- Existe un módulo llamado mod2 el cual contiene la instrucción `import mod1`.
- Hay un archivo principal que contiene las instrucciones `import mod1` e `import mod2`.

A primera vista, se puede pensar que mod1 será importado dos veces - afortunadamente, **solo se produce la primera importación. Python recuerda los módulos importados y omite silenciosamente todas las importaciones posteriores**.

Python puede hacer mucho más. También crea una variable llamada `__name__`.

Además, cada archivo fuente usa su propia versión separada de la variable, no se comparte entre módulos.

Te mostraremos cómo usarlo. Modifica el `module.py` un poco: `print('I like to be a module.\n',__name__)`

Ahora ejecuta el archivo `module.py`. Deberías ver las siguientes líneas:

`I like to be a module.
__main__`

Podemos decir que:

- Cuando se ejecuta un archivo directamente, su variable `__name__` se establece a `__main__`.
- Cuando un archivo se importa como un módulo, su variable `__name__` se establece al nombre del archivo (excluyendo a .py).

Así es como puedes hacer uso de la variable `__main__` para detectar el contexto en el cual se activó tu código: 






In [ ]:
if __name__ == '__main__':
    print('I prefer to be a module.')
else:
    print('I like to be a module.')

Sin embargo, hay una forma más inteligente de utilizar la variable. Si escribes un módulo lleno de varias funciones complejas, puedes usarla para colocar una serie de pruebas para verificar si las funciones trabajan correctamente.

Cada vez que modifiques alguna de estas funciones, simplemente puedes ejecutar el módulo para asegurarte de que sus enmiendas no estropeen el código. Estas pruebas se omitirán cuando el código se importe como un módulo.

Este módulo contendrá dos funciones simples, y si deseas saber cuántas veces se han invocado las funciones, necesitas un contador inicializado en cero cuando se importa el módulo. Puedes hacerlo de esta manera:

In [ ]:
# module.py

counter = 0

if __name__ == "__main__":
    print("I prefer to be a module")
else:
    print("I like to be a module")

El introducir tal variable es absolutamente correcto, pero puede causar importantes efectos secundarios que debes tener en cuenta.

Analiza el archivo modificado `main.py`:

In [ ]:
# main.py

import module

print(module.counter)

Como puedes ver, el archivo principal intenta acceder a la variable de contador del módulo. ¿Es esto legal? Sí lo es. ¿Es utilizable? Claro. ¿Es seguro? Eso depende: si confías en los usuarios de tu módulo, no hay problema; sin embargo, es posible que no desees que el resto del mundo vea tu 88variable personal o privada88.

A diferencia de muchos otros lenguajes de programación, Python no tiene medios para permitirte ocultar tales variables a los ojos de los usuarios del módulo. Solo puedes informar a tus usuarios que esta es tu variable, que pueden leerla, pero que no deben modificarla bajo ninguna circunstancia.

Esto se hace anteponiendo al nombre de la variable `_` (un guión bajo) o `__` (dos guiones bajos), pero recuerda, es solo un **acuerdo**. Los usuarios de tu módulo pueden obedecerlo o no.

Nosotros por supuesto, lo respetaremos. Ahora pongamos dos funciones en el módulo: evaluarán la suma y el producto de los números recopilados en una lista.

Además, agreguemos algunos adornos allí y eliminemos los restos superfluos:

In [ ]:
# module.py

#!/usr/bin/env python3 

""" module.py - an example of Python module """

__counter = 0

def suml(list):
	global __counter
	__counter += 1
	sum = 0
	for el in list:
		sum += el
	return sum

def prodl(list):
	global __counter	
	__counter += 1
	prod = 1
	for el in list:
		prod *= el
	return prod

if __name__ == "__main__":
	print("I prefer to be a module, but I can do some tests for you")
	l = [i+1 for i in range(5)]
	print(suml(l) == 15)
	print(prodl(l) == 120)

Algunos elementos necesitan explicación:

- La línea que comienza con `#!` desde el punto de vista de Python, es solo un **comentario** debido a que comienza con `#`. Para sistemas operativos Unix y similares a Unix (incluido MacOS), dicha línea **indica al sistema operativo cómo ejecutar el contenido del archivo** (en otras palabras, qué programa debe lanzarse para interpretar el texto). En algunos entornos (especialmente aquellos conectados con servidores web) la ausencia de esa línea causará problemas.
- Una cadena (quizás una multilínea) colocada antes de las instrucciones de cualquier módulo (incluidas las importaciones) se denomina **doc-string**, y debe explicar brevemente el propósito y el contenido del módulo.
- Las funciones definidas dentro del módulo (`suml()` y `prodl()`) están disponibles para ser importadas.
- Se ha utilizado la variable `__name__` para detectar cuándo se ejecuta el archivo de forma independiente.

Ahora es posible usar el nuevo módulo, esta es una forma de hacerlo:

In [ ]:
# main.py

from module import suml, prodl

zeroes = [0 for i in range(5)]
ones = [1 for i in range(5)]
print(suml(zeroes))
print(prodl(ones))

Es hora de hacer este ejemplo más complejo: hemos asumido aquí que el archivo Python principal se encuentra en la misma carpeta o directorio que el módulo que se va a importar.

Realicemos el siguiente experimento mental:

- Estamos utilizando el sistema operativo Windows ® (esta suposición es importante, ya que la forma del nombre del archivo depende de ello).
- El script principal de Python se encuentra en C:\Users\user\py\progs y se llama `main.py`.
- El módulo a importar se encuentra en C:\Users\user\py\modules.

¿Como lidiar con ello?

Para responder a esta pregunta, tenemos que hablar sobre **cómo Python busca módulos**. Hay una variable especial (en realidad una lista) que almacena todas las ubicaciones (carpetas o directorios) que se buscan para encontrar un módulo que ha sido solicitado por la instrucción `import`.

Python examina estas carpetas en el orden en que aparecen en la lista: si el módulo no se puede encontrar en ninguno de estos directorios, la importación falla.

De lo contrario, se tomará en cuenta la primera carpeta que contenga un módulo con el nombre deseado (si alguna de las carpetas restantes contiene un módulo con ese nombre, se ignorará).

La variable se llama `path` (ruta), y es accesible a través del módulo llamado `sys`. Así es como puedes verificar su valor:

Hemos lanzado el código dentro del directorio C:\User\user y obtenemos:

In [ ]:
C:\Users\user
C:\Users\user\AppData\Local\Programs\Python\Python36-32\python36.zip
C:\Users\user\AppData\Local\Programs\Python\Python36-32\DLLs
C:\Users\user\AppData\Local\Programs\Python\Python36-32\lib
C:\Users\user\AppData\Local\Programs\Python\Python36-32
C:\Users\user\AppData\Local\Programs\Python\Python36-32\lib\site-packages

N.B.: la carpeta en la que comienza la ejecución aparece en el primer elemento de la ruta.

Ten en cuenta también que: hay un archivo zip listado como uno de los elementos de la ruta, esto no es un error. Python puede tratar los archivos zip como carpetas ordinarias, esto puede ahorrar mucho almacenamiento.

¿Puedes predecir cómo resolver este problema?

Puedes resolverlo agregando una carpeta que contenga el módulo a la variable de ruta (`path`), es completamente modificable.

Una de las varias soluciones posibles se ve así:

In [ ]:
# main.py

from sys import path

path.append('..\\modules')

import module

zeroes = [0 for i in range(5)]
ones = [1 for i in range(5)]
print(module.suml(zeroes))
print(module.prodl(ones))

N.B.:

- Se ha duplicado la \ dentro del nombre de la carpeta, ¿sabes por qué? Debido a que una diagonal invertida se usa para escapar de otros caracteres, si deseas obtener solo una diagonal invertida, debe escapar.
- Hemos utilizado el nombre relativo de la carpeta: esto funcionará si inicia el archivo main.py directamente desde su carpeta de inicio, y no funcionará si el directorio actual no se ajusta a la ruta relativa; siempre puedes usar una ruta absoluta, como esta: `path.append('C:\\Users\\user\\py\\modules')`
- Hemos usado el método `append()`, la nueva ruta ocupará el último elemento en la lista de rutas; si no te gusta la idea, puedes usar en lugar de ello el método `insert()`.

Imagina que en un futuro no muy lejano, tu y tus socios escriben una gran cantidad de funciones en Python.

Tu equipo decide agrupar las funciones en módulos separados, y este es el resultado final:

...

N.B.: hemos presentado todo el contenido solo para el módulo `omega.py`: supongamos que todos los módulos tienen un aspecto similar (contienen una función denominada `funX`, donde `X` es la primera letra del nombre del módulo).

De repente, alguien se da cuenta de que estos módulos forman su propia jerarquía, por lo que colocarlos a todos en una estructura plana no será una buena idea.

Después de algo de discusión, el equipo llega a la conclusión de que los módulos deben agruparse. Todos los participantes están de acuerdo en que la siguiente estructura de árbol refleja perfectamente las relaciones mutuas entre los módulos:

...

Repasemos esto de abajo hacia arriba:

- El grupo ugly contiene dos módulos: psi y omega.
- El grupo best contiene dos módulos: sigma y tau.
- El grupo good contiene dos módulos: (alpha y beta) y un subgrupo (best).
- El grupo extra contiene dos subgrupos: (good y ugly) y un módulo (iota).

¿Se ve mal? De ninguna manera: analiza la estructura cuidadosamente. Se parece a algo, ¿no?

Parece la estructura de un directorio.

Así es como se ve actualmente la relación entre módulos:

...

Tal estructura es casi un paquete (en el sentido de Python). Carece del detalle fino para ser funcional y operativo. Lo completaremos en un momento.

Si asumes que extra es el nombre de un **recientemente creado paquete** (piensa en él como la **raíz del paquete**), impondrá una regla de nomenclatura que te permitirá nombrar claramente cada entidad del árbol.

Por ejemplo:

- La ubicación de una función llamada `funT()` del paquete tau puede describirse como: `extra.good.best.tau.funT()`
- Una función marcada como: `extra.ugly.psi.funP()` proviene del módulo psi el cual esta almacenado en subpaquete ugly del paquete extra.

Se deben responder dos preguntas:

- ¿Cómo se transforma este árbol (en realidad, un subárbol) en un paquete real de Python (en otras palabras, ¿cómo convence a Python de que dicho árbol no es solo un montón de archivos basura, sino un conjunto de módulos)?
- ¿Dónde se coloca el subárbol para que Python pueda acceder a él?

La primer pregunta tiene una respuesta sorprendente: **los paquetes, como los módulos, pueden requerir inicialización**. La inicialización de un módulo se realiza mediante un código independiente (que no forma parte de ninguna función) ubicado dentro del archivo del módulo. Como un paquete no es un archivo, esta técnica es inútil para inicializar paquetes.

En su lugar, debes usar un truco diferente: Python espera que haya un archivo con un nombre muy exclusivo dentro de la carpeta del paquete: `__init__.py`.

El contenido del archivo se ejecuta cuando se importa cualquiera de los módulos del paquete. **Si no deseas ninguna inicialización especial, puedes dejar el archivo vacío, pero no debes omitirlo**.

La presencia del archivo `__init.py__` finalmente compone el paquete:

...

N.B.: no solo la carpeta raiz puede contener el archivo `__init.py__` - también puedes ponerlo dentro de cualquiera de sus subcarpetas (subpaquetes). Puede ser útil si algunos de los subpaquetes requieren tratamiento individual o un tipo especial de inicialización.

Ahora es el momento de responder la segunda pregunta: la respuesta es simple: **donde quiera**. Solo tienes que asegurarte de que Python conozca la ubicación del paquete. Ya sabes cómo hacer eso.

Estás listo para usar tu primer paquete.

Supongamos que el entorno de trabajo se ve de la siguiente manera:

Hemos preparado un archivo zip que contiene todos los archivos de la rama de paquetes. Puedes descargarlo y usarlo para tus propios experimentos, pero recuerda desempaquetarlo en la carpeta presentada en el esquema, de lo contrario, no será accesible para el código.

`Archivo ZIP Módulos y paquetes.zip`

Continuarás tus experimentos usando el archivo `main2.py`.

Vamos a acceder a la función `funI()` del módulo iota del paquete extra. Nos obliga a usar nombres de paquetes calificados (asocia esto al nombramiento de carpetas y subcarpetas).

Asi es como se hace:

In [ ]:
# main2.py

from sys import path

path.append('..\\packages')

import extra.iota

print(extra.iota.funI())

N.B.:

- Hemos modificado la variable `path` para que sea accesible a Python.
- El `import` no apunta directamente al módulo, pero especifica la ruta completa desde la parte superior del paquete.

El reemplazar `import extra.iota` con `import iota` causará un error.

La siguiente variante también es válida:


In [ ]:
# main2.py

from sys import path

path.append('..\\packages')

from extra.iota import funI

print(funI())

Nota el nombre calificado del módulo iota.

Ahora vamos hasta el final del árbol: así es como se obtiene acceso a los módulos sigma y tau.

In [ ]:
from sys import path

path.append('..\\packages')

import extra.good.best.sigma
from extra.good.best.tau import funT

print(extra.good.best.sigma.funS())
print(funT())

Puedes hacer tu vida más fácil usando un alias:

In [ ]:
# main2.py

from sys import path

path.append('..\\packages')

import extra.good.best.sigma as sig
import extra.good.alpha as alp

print(sig.funS())
print(alp.funA())

Supongamos que hemos comprimido todo el subdirectorio, comenzando desde la carpeta extra (incluyéndola), y se obtuvo un archivo llamado `extrapack.zip`. Después, colocamos el archivo dentro de la carpeta packages.

Ahora podemos usar el archivo zip en un rol de paquetes:

In [ ]:
# main2.py

from sys import path

path.append('..\\packages\\extrapack.zip')

import extra.good.best.sigma as sig
import extra.good.alpha as alp
from extra.iota import funI
from extra.good.beta import funB

print(sig.funS())
print(alp.funA())
print(funI())
print(funB())

Si deseas realizar tus propios experimentos con el paquete que hemos creado, puedes descargarlo a continuación. Te alentamos a que lo hagas.

Ahora puedes crear módulos y combinarlos en paquetes. Es hora de comenzar una discusión completamente diferente: sobre errores y fallas.

## 4.2 Manejo de excepciones

### 4.2.1 Errores: el pan diario del programador

**Cualquier cosa que pueda salir mal, saldrá mal**.

Esta es la ley de Murphy, y funciona en todo y siempre. Si la ejecución del código puede salir mal, lo hará.

Observa el código en el editor. Hay al menos dos formas posibles de que "salga mal" la ejecución. ¿Puedes verlas?

In [ ]:
import math

x = float(input("Ingresa x: "))
y = math.sqrt(x)

print("La raíz cuadrada de", x, "es igual a", y)

Ingresa x: -4


ValueError: ignored

- Como el usuario puede ingresar una cadena de caracteres completamente arbitraria, **no hay garantía de que la cadena se pueda convertir en un valor flotante** - esta es la primera vulnerabilidad del código.
- La segunda es que la función `sqrt()` **fallará si se le ingresa un valor negativo**.

Puedes recibir alguno de los siguientes mensajes de error:

In [ ]:
Ingresa x: Abracadabra

Traceback (most recent call last):

  File "sqrt.py", line 3, in 

    x = float(input("Ingresa x: "))

ValueError: could not convert string to float: 'Abracadabra'

In [ ]:
Ingresa x: -1

Traceback (most recent call last):

  File "sqrt.py", line 4, in 

    y = math.sqrt(x)

ValueError: math domain error

¿Puedes protegerte de tales sorpresas? Por supuesto que si. Además, tienes que hacerlo para ser considerado un buen programador.

Cada vez que tu código intenta hacer algo erroneo, irresponsable o inaplicable, Python hace dos cosas:

- **Detiene tu programa**.
- Crea un tipo especial de dato, llamado **excepción**.

Ambas actividades llevan por nombre **lanzar una excepción**. Podemos decir que Python siempre lanza una excepción (o que **una excepción ha sido lanzada**) cuando no tiene idea de qué hacer con el código.

¿Qué ocurre después?

- La excepción lanzada espera que alguien o algo lo note y haga algo al respecto.
- **Si la excepción no es resuelta, el programa será terminado abruptamente, y verás un mensaje de error** enviado a la consola por Python.
- De otra manera, si se atiende la excepción y es **manejada** apropiadamente, el programa puede reanudarse y su ejecución puede continuar.

Python proporciona herramientas efectivas que permiten **observar, identificar y manejar las excepciones** eficientemente. Esto es posible debido a que todas las excepciones potenciales tienen un nombre específico, por lo que se pueden clasificar y reaccionar a ellas adecuadamente.

Ya conoces algunos nombres de excepción. Observa el siguiente mensaje de diagnóstico: `ValueError: math domain error` 

La palabra en rojo (`ValueError`) es solo el **nombre de la excepción**. 

¿Cómo se **manejan** las excepciones? La palabra `try` es clave para la solución.

Además, también es una palabra reservada.

La receta para el éxito es la siguiente:

- Primero, se debe **intentar (try) hacer algo**.
- Después, tienes que **comprobar si todo salió bien**.

Pero, ¿no sería mejor verificar primero todas las circunstancias y luego hacer algo solo si es seguro?

In [ ]:
primerNumero = int(input("Ingresa el primer numero: "))
segundoNumero = int(input("Ingresa el segundo numero: "))

if segundoNumero != 0:
    print(primerNumero / segundoNumero)
else:
    print("Esta operacion no puede ser realizada.")

print("FIN.")

Ingresa el primer numero: 1
Ingresa el segundo numero: 0
Esta operacion no puede ser realizada.
FIN.


Es cierto que esta forma puede parecer la más natural y comprensible, pero en realidad, este método no facilita la programación. Todas estas revisiones pueden hacer el código demasiado grande e ilegible. Python prefiere un enfoque completamente diferente:

- La palabra reservada `try` **comienza con un bloque de código** el cual puede o no estar funcionando correctamente.
- Después, Python intenta realizar la acción arriesgada: si falla, se genera una excepción y Python comienza a buscar una solución.
- La palabra reservada `except` comienza con un bloque de código que será **ejecutado si algo dentro del bloque** `try` **sale mal** - si se genera una excepción dentro del bloque anterior `try`, **fallará aquí**, entonces el código ubicado después de la palabra clave `except` debería proporcionar una **reacción adecuada** a la excepción planteada.
- Se regresa al nivel de anidación anterior, es decir, se termina la sección **try-except**.

In [ ]:
try:
    print("1")
    x = 1 / 0
    print("2")
except:
    print("Oh cielos, algo salio mal...")

print("3")

1
Oh cielos, algo salio mal...
3


Este enfoque tiene una desventaja importante: si existe la posibilidad de que más de una excepción se salte a un apartado `except:`, puedes tener **problemas para descubrir lo que realmente sucedió**.

In [ ]:
try:
    x = int(input("Ingresa un numero: "))
    y = 1 / x
except:
    print("Oh cielos, algo salio mal...")

print("FIN.")

Ingresa un numero: 0
Oh cielos, algo salio mal...
FIN.


El mensaje: `Oh cielos, algo salio mal...` que aparece en la consola no dice nada acerca de la razón, mientras que hay dos posibles causas de la excepción:

- Datos no enteros fueron ingresados por el usuario.
- Un valor entero igual a `0` fue asignado a la variable `x`.

Técnicamente, hay dos formas de resolver el problema:

- Construir dos bloques consecutivos try-except, uno por cada posible motivo de excepción (fácil, pero provocará un crecimiento desfavorable del código).
- Emplear una variante más avanzada de la instrucción.

In [ ]:
try:
    :
except exc1:
    :
except exc2:
    :
except:
    :

Así es como funciona:

- Si el `try` lanza la excepción `exc1`, esta será manejada por el bloque `except exc1:`.
- De la misma manera, si el `try` lanza la excepción `exc2`, esta será manejada por el bloque `except exc2:`.
- Si el `try` lanza cualquier otra excepción, será manejado por el bloque sin nombre `except:`.

In [ ]:
try:
    x = int(input("Ingresa un numero: "))
    y = 1 / x
    print(y)
except ZeroDivisionError:
    print("No puedes dividir entre cero, lo siento.")
except ValueError:
    print("Debes ingresar un valor entero.")
except:
    print("Oh cielos, algo salio mal...")

print("THE END.")

Oh cielos, algo salio mal...
THE END.


El código, cuando se ejecute, producirá una de las siguientes cuatro variantes de salida:

- Si se ingresa un valor entero válido distinto de cero (por ejemplo, `5`) dirá:
`0.2 FIN.`

- Si se ingresa `0`, dirá: `No puedes dividir entre cero, lo siento.
FIN.`
`
- Si se ingresa cualquier cadena no entera, verás: `Debes ingresar un valor entero. FIN.`

- (Localmente en tu máquina) si presionas Ctrl-C mientras el programa está esperando la entrada del usuario (provocará una excepción denominada `KeyboardInterrupt`), el programa dirá: `Oh cielos, algo salio mal...
FIN.`

No olvides que:

- Los bloques `except` son analizados en el mismo orden en que aparecen en el código.
- No debes usar más de un bloque de excepción con el mismo nombre.
- El número de diferentes bloques 1except1 es arbitrario, la única condición es que si se emplea el `try`, debes poner al menos un `except` (nombrado o no) después de el.
- La palabra reservada `except` no debe ser empleada sin que le preceda un `try`.
- Si uno de los bloques `except` es ejecutado, ningún otro lo será.
- Si ninguno de los bloques `except` especificados coincide con la excepción planteada, la excepción permanece sin manejar (lo discutiremos pronto).
- Si un `except` sin nombre existe, tiene que especificarse como el último.


In [ ]:
try:
    x = int(input("Ingresa un numero: "))
    y = 1 / x
    print(y)
except ValueError:
    print("Debes ingresar un valor entero.")

print("FIN.")

Ingresa un numero: 0


ZeroDivisionError: ignored

En la siguiente sección, nos centraremos en las excepciones integradas de Python y sus jerarquías.

### 4.2.2 La anatomía de las excepciones

Python 3 define **63 excepciones integradas**, y todos ellos forman una **jerarquía en forma de árbol**, aunque el árbol es un poco extraño ya que su raíz se encuentra en la parte superior.

Algunas de las excepciones integradas son más generales (incluyen otras excepciones) mientras que otras son completamente concretas (solo se representan a sí mismas). Podemos decir que **cuanto más cerca de la raíz se encuentra una excepción, más general (abstracta) es**. A su vez, las excepciones ubicadas en los extremos del árbol (podemos llamarlas **hojas**) son **concretas**.

Echa un vistazo a la figura:

...

Muestra una pequeña sección del árbol completo de excepciones. Comencemos examinando el árbol desde la hoja `ZeroDivisionError`:

- `ZeroDivisionError` es un caso especial de una clase de excepción más general llamada `ArithmeticError`.
- `ArithmeticError` es un caso especial de una clase de excepción más general llamada solo `Exception`.
- `Exception` es un caso especial de una clase más general llamada `BaseException`.

Podemos describirlo de la siguiente manera (observa la dirección de las flechas; siempre apuntan a la entidad más general):

...


In [ ]:
try:
    y = 1 / 0
except ZeroDivisionError:
    print("Uuuppsss...")

print("FIN.")

Uuuppsss...
FIN.


In [ ]:
try:
    y = 1 / 0
except ArithmeticError:
    print("Uuuppsss...")

print("FIN.")

Por lo tanto, la salida del código permanece sin cambios. Esto también significa que reemplazar el nombre de la excepción ya sea con `Exception` o `BaseException` no cambiará el comportamiento del programa.

In [ ]:
try:
    y = 1 / 0
except ZeroDivisionError:
    print("¡División entre Cero!")
except ArithmeticError:
    print("¡Problema aritmético!")

print("FIN.")

¡División entre Cero!
FIN.


In [ ]:
try:
    y = 1 / 0
except ArithmeticError:
    print("¡Problema aritmético!")
except ZeroDivisionError:
    print("¡División entre Cero!")

print("FIN.")

¡Problema aritmético!
FIN.


¿Por qué, si la excepción planteada es la misma que antes?

La excepción es la misma, pero la excepción más general ahora aparece primero: también capturará todas las divisiones entre cero. También significa que no hay posibilidad de que alguna excepción llegue a `ZeroDivisionError`. Ahora es completamente inalcanzable.

Recuerda:

- ¡El orden de las excepciones importa!
- No pongas excepciones más generales antes que otras más concretas.
- Esto hará que el último sea inalcanzable e inútil.
- Además, hará que el código sea desordenado e inconsistente.
- Python no generará ningún mensaje de error con respecto a este problema.

Si deseas manejar dos o mas excepciones de la misma manera, puedes usar la siguiente sintaxis:


In [ ]:
try:
    :
except (exc1, exc2):
    :

Simplemente tienes que poner todos los nombres de excepción empleados en una lista separada por comas y no olvidar los paréntesis.

**Si una excepción se genera dentro de una función**, puede ser manejada:

- Dentro de la función.
- Fuera de la función.

#### Excepción manejada dentro de la función

Comencemos con la primera variante: 


In [ ]:
def badFun(n):
    try:
        return 1 / n
    except ArithmeticError:
        print("¡Problema aritmético!")
    return None

badFun(0)

print("FIN.")

¡Problema aritmético!
FIN.


La excepción `ZeroDivisionError` (la cual es un caso concreto de la clase `ArithmeticError`) es lanzada dentro de la función `badfun()`, y la función en sí misma se encarga de su caso.

#### Excepción manejada fuera de la función

La instrucción `raise` genera la excepción especificada denominada `exc` como si fuese generada de manera natural: `raise exc`

N.B.: `raise` es una palabra reservada.

La instrucción permite:

- **Simular excepciones reales** (por ejemplo, para probar tu estrategia de manejo de excepciones).
- Parcialmente **manejar una excepción** y hacer que otra parte del código sea responsable de completar el manejo.

In [ ]:
def badFun(n):
    raise ZeroDivisionError

try:
    badFun(2)
except ArithmeticError:
    print("¿Que pasó? ¿Un error?")

print("FIN.")

¿Que pasó? ¿Un error?
FIN.


La salida del programa permanece sin cambios. De esta manera, puedes **probar tu rutina de manejo de excepciones** sin forzar al código a hacer cosas incorrectas.

La instrucción `raise` también se puede utilizar de la siguiente manera (toma en cuenta la ausencia del nombre de la excepción): `raise`

Existe una seria restricción: esta variante de la instrucción `raise` puede ser utilizada **solamente dentro de la rama** `except`; usarla en cualquier otro contexto causa un error.

La instrucción volverá a generar la misma excepción que se maneja actualmente. Gracias a esto, puedes distribuir el manejo de excepciones entre diferentes partes del código.

In [ ]:
def badFun(n):
    try:
        return n / 0
    except:
        print("¡Lo hice otra vez!")
        raise 

try:
    badFun(0)
except ArithmeticError:
    print("¡Ya veo!")

print("FIN.")

¡Lo hice otra vez!
¡Ya veo!
FIN.


La excepción `ZeroDivisionError` es generada dos veces:

- Primero, dentro del `try` debido a que se intentó realizar una división entre cero.
- Segundo, dentro de la parte `except` por la instrucción `raise`.

Ahora es un buen momento para mostrarte otra instrucción de Python, llamada `assert` (afirmar). Esta es una palabra reservada. `assert expresión`

¿Como funciona?

- Evalúa la `expresión`.
- Si la expresión se evalúa como `True` (verdadero), o un valor numérico distinto de cero, o una cadena no vacía, o cualquier otro valor diferente de `None`, no hará nada más.
- De lo contrario, automáticamente e inmediatamente genera una excepción llamada `AssertionError` (en este caso, decimos que la afirmación ha fallado).
¿Cómo puede ser utilizada?

- Puedes ponerlo en la parte del código donde quieras estar **absolutamente a salvo de datos incorrectos**, y donde no estés absolutamente seguro de que los datos hayan sido examinados cuidadosamente antes (por ejemplo, dentro de una función utilizada por otra persona).
- El generar una excepción `AssertionError` asegura que tu código no produzca resultados no válidos y muestra claramente la naturaleza de la falla.
- **Las aserciones no reemplazan las excepciones ni validan los datos**, son suplementos.

Si las excepciones y la validación de datos son como conducir con cuidado, la aserción puede desempeñar el papel de una bolsa de aire.

In [ ]:
import math

x = float(input("Ingresa un numero: "))
assert x >= 0.0

x = math.sqrt(x)

print(x)

Ingresa un numero: -1


AssertionError: ignored

#### Excepciones integradas

Te mostraremos una breve lista de las excepciones más útiles. Si bien puede sonar extraño llamar "útil" a una cosa o un fenómeno que es un signo visible de una falla o retroceso, como sabes, errar es humano y si algo puede salir mal, saldrá mal.

Las excepciones son tan rutinarias y normales como cualquier otro aspecto de la vida de un programador.

Para cada excepción, te mostraremos:

- Su nombre.
- Su ubicación en el árbol de excepciones.
- Una breve descripción.
- Un fragmento de código conciso que muestre las circunstancias en las que se puede generar la excepción.

Hay muchas otras excepciones para explorar: simplemente no tenemos el espacio para revisarlas todas aquí.

...

Hemos terminado con excepciones por ahora, pero volverán cuando discutamos la programación orientada a objetos en Python. Puedes usarlos para proteger tu código de accidentes graves, pero también tienes que aprender a sumergirte en ellos, explorando la información que llevan.

De hecho, las excepciones son objetos; sin embargo, no podemos decirle nada sobre este aspecto hasta que te presentemos clases, objetos y similares.

Por el momento, si deseas obtener más información sobre las excepciones por tu cuenta, consulta la Biblioteca estándar de Python en https://docs.python.org/3.6/library/exceptions.html.


In [ ]:
try:
    val = int(input('ingrese un número entero:'))
    assert val >= 0
except ValueError:
    print('¡Debe ingresar solo números enteros!')
    print('Hubo un error al convertir texto en entero.')
except AssertionError:
    print('¡Debe ingresar enteros >=0!')
except:
    print('Mucho cuidado.')

try:
    num = 10/val
except ZeroDivisionError:
    print('Hubo un error al intentar dividir entre cero.')
except:
    print('Mucho cuidado...')
print('num\t:\t',num)

ingrese un número entero:3
num	:	 3.3333333333333335


## 4.3 Cadenas

**Las computadoras almacenan los caracteres como números**. Cada carácter utilizado por una computadora corresponde a un número único, y viceversa. Esta asignación debe incluir más caracteres de los que podrías esperar. Muchos de ellos son invisibles para los humanos, pero esenciales para las computadoras.

Algunos de estos caracteres se llaman **espacios en blanco**, mientras que otros se nombran **caracteres de control**, porque su propósito es controlar dispositivos de entrada/salida.

Un ejemplo de un espacio en blanco que es completamente invisible a simple vista es un código especial, o un par de códigos (diferentes sistemas operativos pueden tratar este asunto de manera diferente), que se utilizan para marcar el final de las líneas dentro de los archivos de texto.

Las personas no ven este signo (o estos signos), pero pueden observar el efecto de su aplicación donde ven un salto de línea.

Podemos crear prácticamente cualquier cantidad de asignaciones de números con caracteres, pero la vida en un mundo en el que cada tipo de computadora utiliza una codificación de caracteres diferentes no sería muy conveniente. Este sistema ha llevado a la necesidad de introducir un estándar universal y ampliamente aceptado, implementado por (casi) todas las computadoras y sistemas operativos en todo el mundo.

El denominado **ASCII** (por sus siglas en íngles American Standard Code for Information Interchange). El Código Estándar Americano para Intercambio de Información es el más utilizado, y es posible suponer que casi todos los dispositivos modernos (como computadoras, impresoras, teléfonos móviles, tabletas, etc.) usan este código.

El código proporciona espacio para **256** caracteres diferentes, pero solo nos interesan los primeros 128. Si deseas ver cómo se construye el código, mira la tabla a continuación. Haz clic en la tabla para ampliarla. Mírala cuidadosamente: hay algunos datos interesantes. Observa el código del caracter más común: el espacio. El cual es el 32.

### 5.3.1 Puntos de código y páginas de códigos

Necesitamos un nuevo término: un **punto de código**.

Un punto de código es un **numero que compone un caracter**. Por ejemplo, 32 es un punto de código que compone un espacio en codificación ASCII. Podemos decir que el código ASCII estándar consta de 128 puntos de código.

Como el ASCII estándar ocupa 128 de 256 puntos de código posibles, solo puedes hacer uso de los 128 restantes.

No es suficiente para todos los idiomas posibles, pero puede ser suficiente para un idioma o para un pequeño grupo de idiomas similares.

¿Se puede **establecer la mitad superior de los puntos de código de manera diferente para diferentes idiomas**? Si, por supuesto. A tal concepto se le denomina una **página de códigos**.

Una página de códigos es un **estándar para usar los 128 puntos de código superiores para almacenar caracteres específicos**. Por ejemplo, hay diferentes páginas de códigos para Europa Occidental y Europa del Este, alfabetos cirílicos y griegos, idiomas árabe y hebreo, etc.

Esto significa que el mismo punto de código puede formar diferentes caracteres cuando se usa en diferentes páginas de códigos.

Por ejemplo, el punto de código 200 forma una Č (una letra usada por algunas lenguas eslavas) cuando lo utiliza la página de códigos ISO/IEC 8859-2, pero forma un Ш (una letra cirílica) cuando es usado por la página de códigos ISO/IEC 8859-5.

En consecuencia, para determinar el significado de un punto de código específico, debes conocer la página de códigos de destino.

En otras palabras, los puntos de código derivados del código de páginas son ambiguos

#### Unicode

Las páginas de códigos ayudaron a la industria informática a resolver problemas de I18N durante algún tiempo, pero pronto resultó que no serían una solución permanente. El concepto que resolvió el problema a largo plazo fue el **Unicode**.

**Unicode asigna caracteres únicos (letras, guiones, ideogramas, etc.) a más de un millón de puntos de código**. Los primeros 128 puntos de código Unicode son idénticos a ASCII, y los primeros 256 puntos de código Unicode son idénticos a la página de códigos ISO / IEC 8859-1 (una página de códigos diseñada para idiomas de Europa occidental).

#### UCS-4

El estándar Unicode no dice nada sobre cómo codificar y almacenar los caracteres en la memoria y los archivos. Solo nombra todos los caracteres disponibles y los asigna a planos (un grupo de caracteres de origen, aplicación o naturaleza similares).

Existe más de un estándar que describe las técnicas utilizadas para implementar Unicode en computadoras y sistemas de almacenamiento informáticos reales. El más general de ellos es **UCS-4**. El nombre viene de **Universal Character Set** (Conjunto de Caracteres Universales).

**UCS-4 emplea 32 bits (cuatro bytes) para almacenar cada caracter**, y el código es solo el número único de los puntos de código Unicode. Un archivo que contiene texto codificado UCS-4 puede comenzar con un BOM (byte order mark - marca de orden de bytes), una combinación no imprimible de bits que anuncia la naturaleza del contenido del archivo. Algunas utilidades pueden requerirlo.

Como puedes ver, UCS-4 es un estándar bastante derrochador: aumenta el tamaño de un texto cuatro veces en comparación con el estándar ASCII. Afortunadamente, hay formas más inteligentes de codificar textos Unicode.

#### UTF-8

Uno de los más utilizados es **UTF-8**. El nombre se deriva de **Unicode Transformation Format** (Formato de Transformación Unicode).

El concepto es muy inteligente. **UTF-8 emplea tantos bits para cada uno de los puntos de código como realmente necesita para representarlos**.

Por ejemplo:

- Todos los caracteres latinos (y todos los caracteres ASCII estándar) ocupan ocho bits.
- Los caracteres no latinos ocupan 16 bits.
- Los ideógrafos CJK (China-Japón-Corea) ocupan 24 bits.

Debido a las características del método utilizado por UTF-8 para almacenar los puntos de código, no es necesario usar el BOM, pero algunas de las herramientas lo buscan al leer el archivo, y muchos editores lo configuran durante el guardado.

**Python 3 es totalmente compatible con Unicode y UTF-8**:

- Puedes usar caracteres codificados Unicode / UTF-8 para nombrar variables y otras entidades.
- Puedes usarlos durante todas las entradas y salidas.
Esto significa que **Python 3 está completamente Internacionalizado**.


### 4.3.2 La naturaleza de las cadenas en Python

En primer lugar, las cadenas de Python (o simplemente cadenas, ya que no vamos a discutir las cadenas de ningún otro lenguaje) son **secuencias inmutables**.

Es muy importante tener en cuenta esto, porque significa que debes esperar un comportamiento familiar.

Por ejemplo, la función `len()` empleada por cadenas devuelve el número de caracteres que contiene el argumento.

No olvides que la diagonal invertida (\) empleada como un caracter de escape, no esta incluida en la longitud total de la cadena.

In [ ]:
yo_soy = 'I\'m'
print(len(yo_soy))
print(yo_soy)

3
I'm


#### Cadenas multilínea

Ahora es un muy buen momento para mostrarte otra forma de especificar cadenas dentro del código fuente de Python. Ten en cuenta que la sintaxis que ya conoces no te permitirá usar una cadena que ocupe más de una línea de texto.

Por esta razón, el código aquí es erróneo:

In [ ]:
multiLinea = 'Linea #1
Linea #2'

print(len(multiLinea))

SyntaxError: ignored

Afortunadamente, para este tipo de cadenas, Python ofrece una sintaxis simple, conveniente y separada.

In [ ]:
multiLinea = '''Linea #1

Linea #2'''

print(len(multiLinea))
print(multiLinea)

18
Linea #1

Linea #2


Como puedes ver, la cadena comienza con tres apóstrofes, no uno. El mismo apóstrofe triplicado se usa para terminar la cadena. El número de líneas de texto dentro de una cadena de este tipo es arbitrario.

La `Linea #1` contiene ocho caracteres. Las dos líneas juntas contienen 16 caracteres. ¿Perdimos un caracter? ¿Dónde? ¿Cómo?

El caracter que falta es simplemente invisible: es un espacio en blanco. Se encuentra entre las dos líneas de texto. Se denota como: `\n`. ¿Lo recuerdas? Es un caracter especial (de control) utilizado para forzar un avance de línea. No puedes verlo, pero cuenta.

Las cadenas multilínea pueden ser delimitadas también por comillas triples, como aqui:

In [ ]:
multiLinea = """Linea #1
Linea #2"""

print(len(multiLinea))

17


#### Operaciones con Cadenas

Al igual que otros tipos de datos, las cadenas tienen su propio conjunto de operaciones permitidas, aunque son bastante limitadas en comparación con los números. En general, las cadenas pueden ser:

- **Concatenadas** (unidas).
- **Replicadas**.

La primera operación la realiza el operador `+` (toma en cuenta que no es una adición o suma) mientras que la segunda por el operador `*` (toma en cuenta de nuevo que no es una multiplicación).

La capacidad de usar el mismo operador en tipos de datos completamente diferentes (como números o cadenas) se llama **overloading** - sobrecarga (debido a que el operador está sobrecargado con diferentes tareas).

Analiza el ejemplo:

- El operador `+` es empleado en dos o más cadenas y produce una nueva cadena que contiene todos los caracteres de sus argumentos (nota: el orden es relevante aquí, en contraste con su versión numérica, la cual es conmutativa).
- El operador `*` necesita una cadena y un número como argumentos; en este caso, el orden no importa: puedes poner el número antes de la cadena, o viceversa, el resultado será el mismo: una nueva cadena creada por la enésima replicación de la cadena del argumento.

In [ ]:
str1 = 'a'
str2 = 'b'

print(str1 + str2)
print(str2 + str1)
print(5 * 'a')
print('b' * 4)

ab
ba
aaaaa
bbbb


Si deseas saber el valor del punto de código ASCII/UNICODE de un caracter específico, puedes usar la función `ord()` (proveniente de ordinal).

La función necesita una cadena de un caracter como argumento - incumplir este requisito provoca una excepción `TypeError`, y devuelve un número que representa el punto de código del argumento.

In [ ]:
# Demostrando la función ord ()

ch1 = '@' 
ch2 = ' ' # espacio

print(ord(ch1))
print(ord(ch2))
print('type(ch1)',type(ch1))

64
32
type(ch1) <class 'str'>


Si conoces el punto de código (número) y deseas obtener el carácter correspondiente, puedes usar la función llamada `chr()`. La función **toma un punto de código y devuelve su carácter**.

Invocándolo con un argumento inválido (por ejemplo, un punto de código negativo o inválido) provoca las excepciones `ValueError` o `TypeError`.

In [ ]:
# Demostrando la función chr()

print(chr(97))
print(chr(945))

a
α


Ya dijimos antes que las **cadenas de Python son secuencias**. Es hora de mostrarte lo que significa realmente.

**Las cadenas no son listas, pero pueden ser tratadas como tal en muchos casos**. Por ejemplo, si deseas acceder a cualquiera de los caracteres de una cadena, puedes hacerlo usando **indexación**, al igual que en el ejemplo en el editor.

In [ ]:
# Indexando cadenas

exampleString = 'silly walks'

for ix in range(len(exampleString)):
    print(exampleString[ix], end=' ')

s i l l y   w a l k s 

Ten cuidado, no intentes pasar los límites de la cadena, ya que provocará una excepción. Por cierto, los índices negativos también se comportan como se esperaba. Comprueba esto tú mismo.

**Iterar a través de las cadenas** funciona también. 

In [ ]:
# Iterando a través de una cadena

exampleString = 'silly walks'

for ch in exampleString:
    print(ch, end=' ')

s i l l y   w a l k s 

Todo lo que sabes sobre **rodajas** o **rebanadas** es utilizable. Hemos reunido algunos ejemplos que muestran cómo funcionan las rodajas en el mundo de las cadenas.

In [ ]:
# Rodajas o rebanadas

alpha = "abdefg"

print(alpha[1:3])
print(alpha[3:])
print(alpha[:3])
print(alpha[3:-2])
print(alpha[-3:4])
print(alpha[::2])
print(alpha[1::2])

bd
efg
abd
e
e
adf
beg


El operador `in` no debería sorprenderte cuando se aplica a cadenas, simplemente comprueba si el argumento izquierdo (una cadena) se puede encontrar en cualquier lugar dentro del argumento derecho (otra cadena).

El resultado es simplemente `True` (Verdadero) o `False` (Falso).

In [ ]:
alpfabeto = "abcdefghijklmnopqrstuvwxyz"

print("f" in alpfabeto)
print("F" in alpfabeto)
print("1" in alpfabeto)
print("ghi" in alpfabeto)
print("Xyz" in alpfabeto)

True
False
False
True
False


Como probablemente puedas deducir, el operador `not in` también es aplicable aquí.

In [ ]:
alfabeto = "abcdefghijklmnopqrstuvwxyz"

print("f" not in alfabeto)
print("F" not in alfabeto)
print("1" not in alfabeto)
print("ghi" not in alfabeto)
print("Xyz" not in alfabeto)

También te hemos dicho que las **cadenas de Python son inmutables**. Esta es una característica muy importante. ¿Qué significa?

Esto significa principalmente que la similitud de cadenas y listas es limitada. No todo lo que puede hacerse con una lista puede hacerse con una cadena.

La primera diferencia importante no te permite usar la instrucción `del` para eliminar cualquier cosa de una cadena. El ejemplo siguiente no funcionará:

In [ ]:
alfabeto = "abcdefghijklmnopqrstuvwxyz"

del alfabeto[0]

TypeError: ignored

Lo único que puedes hacer con `del` y una cadena es eliminar la cadena como un todo. Intenta hacerlo.

Las cadenas de Python no tienen el método `append()` - no se pueden expander de ninguna manera.

El siguiente ejemplo es erróneo:

In [ ]:
alfabeto = "abcdefghijklmnopqrstuvwxyz"

alfabeto.append("A")

AttributeError: ignored

In [ ]:
msj = 'Hola'
print('id(msj)',id(msj))
msj = msj + ' mundo!'
print('id(msj)',id(msj))

id(msj) 139716812291632
id(msj) 139716811160368


No pienses que la inmutabilidad de una cadena limita tu capacidad de operar con ellas. La única consecuencia es que debes recordarlo e implementar tu código de una manera ligeramente diferente:

In [ ]:
alfabeto = "bcdefghijklmnopqrstuvwxy"

alfabeto = "a" + alfabeto
alfabeto = alfabeto + "z"

print(alfabeto)

Es posible que desees preguntar si el **crear una nueva copia de una cadena cada vez que se modifica su contenido empeora la efectividad del código**. Sí lo hace. Un poco. Sin embargo, no es un problema en absoluto.

Ahora que comprendes que las cadenas son secuencias, podemos mostrarte algunas capacidades de secuencia menos obvias. Las presentaremos utilizando cadenas, pero no olvides que las listas también pueden adoptar los mismos trucos.

Comenzaremos con la función llamada `min()`.

Esta función **encuentra el elemento mínimo de la secuencia pasada como argumento**. Existe una condición - la secuencia (cadena o lista) **no puede estar vacía**, de lo contrario obtendrás una excepción ValueError.

In [ ]:
# Demonstrando min() - Ejemplo 1
print(min("aAbByYzZ"))


# Demonstrando min() - Examplos 2 y 3
t = 'Los Caballeros Que Dicen "¡Ni!"'
print('[' + min(t) + ']')

t = [0, 1, 2]
print(min(t))

A
[ ]
0


Del mismo modo, una función llamada `max()` encuentra el elemento máximo de la secuencia.

In [ ]:
# Demostrando max() - Ejemplo 1
print(max("aAbByYzZ"))


# Demonstrando max() - Examplos 2 y 3
t = 'Los Caballeros Que Dicen "¡Ni!"'
print('[' + max(t) + ']')

t = [0, 1, 2]
print(max(t))

z
[¡]
2


El método `index()` (es un método, no una función) **busca la secuencia desde el principio, para encontrar el primer elemento del valor especificado en su argumento**.

N.B.: el elemento buscado debe aparecer en la secuencia - su ausencia causará una excepción `ValueError`.

El método devuelve el **índice de la primera aparición del argumento** (lo que significa que el resultado más bajo posible es 0, mientras que el más alto es la longitud del argumento decrementado por 1).

In [ ]:
# Demonstrando el método index()
print("aAbByYzZaA".index("b"))
print("aAbByYzZaA".index("Z"))
print("aAbByYzZaA".index("A"))

2
7
1


La función `list()` **toma su argumento (una cadena) y crea una nueva lista que contiene todos los caracteres de la cadena, uno por elemento de la lista**.

N.B.: no es estrictamente una función de cadenas - `list()` es capaz de crear una nueva lista de muchas otras entidades (por ejemplo, de tuplas y diccionarios).

El método `count()` **cuenta todas las apariciones del elemento dentro de la secuencia**. La ausencia de tal elemento no causa ningún problema.

In [ ]:
# Demostrando la función list()
print(list("abcabc"))

# Demostrando el método count()
print("abcabc".count("b"))
print('abcabc'.count("d"))

['a', 'b', 'c', 'a', 'b', 'c']
2
0


Las cadenas de Python tienen un número significativo de métodos destinados exclusivamente al procesamiento de caracteres. No esperes que trabajen con otras colecciones. La lista completa se presenta aquí: https://docs.python.org/3.4/library/stdtypes.html#string-methods.

Te mostraremos los que consideramos más útiles.

In [ ]:
# Demostración del método capitalize()
print('aBcD'.capitalize())

Abcd


In [ ]:
# Demostración del método center()
print('[' + 'alfa'.center(10) + ']')

[   alfa   ]


In [ ]:
# Demostración del método endswith()
if "epsilon".endswith("on"):
    print("si")
else:
    print("no")

si


El método `find()` es similar al método `index()`, el cual ya conoces - busca una subcadena y devuelve el índice de la primera aparición de esta subcadena, pero:

- Es más seguro, **no genera un error para un argumento que contiene una subcadena inexistente** (devuelve `-1` en dicho caso).
- Funciona solo con cadenas - no intentes aplicarlo a ninguna otra secuencia.

In [ ]:
# Demostración del método find()
print("Eta".find("ta"))
print("Eta".find("mma"))

1
-1


In [ ]:
# Demostración del método the isalnum()
print('lambda30'.isalnum())
print('lambda'.isalnum())
print('30'.isalnum())
print('@'.isalnum())
print('lambda_30'.isalnum())
print(''.isalnum())

True
True
True
False
False
False


In [ ]:
t = 'Six lambdas'
print(t.isalnum())

t = 'ΑβΓδ'
print(t.isalnum())

t = '20E1'
print(t.isalnum())

False
True
True


In [ ]:
# Ejemplo 1: Demostración del método isapha()
print("Moooo".isalpha())
print('Mu40'.isalpha())

# Ejemplo 2: Demostración del método isdigit()
    
print('2018'.isdigit())
print("Año2019".isdigit())

True
False
True
False


In [ ]:
# Ejemplo 1: Demostración del método islower()
print("Moooo".islower())
print('moooo'.islower())
print()

# Ejemplo 2: Demostración del método isspace()
print(' \n '.isspace())
print(" ".isspace())
print("mooo mooo mooo".isspace())
print()

# Ejemplo 3: Demostración del método isupper() 
print("Moooo".isupper())
print('moooo'.isupper())
print('MOOOO'.isupper())

False
True

True
True
False

False
False
True


El método `join()` es algo complicado, así que déjanos guiarte paso a paso:

- Como su nombre lo indica, el método **realiza una unión** y espera un argumento del tipo lista; se debe asegurar que todos los elementos de la lista sean cadenas: de lo contrario, el método generará una excepción `TypeError`.
- Todos los elementos de la lista **serán unidos en una sola cadena** pero...
- ...la cadena desde la que se ha invocado el método será **utilizada como separador**, puesta entre las cadenas.
- La cadena recién creada se devuelve como resultado.

In [ ]:
# Demostración del método join()
print(",".join(["omicron", "pi", "rho"]))

omicron,pi,rho


Vamos a analizarlo:
- El método `join()` se invoca desde una cadena que contiene una coma (la cadena puede ser larga o puede estar vacía).
- El argumento del `join` es una lista que contiene tres cadenas.
- El método devuelve una nueva cadena.

In [ ]:
# Demostración del método lower()
print("SiGmA=60".lower())

sigma=60


El método sin parámetros `lstrip()` devuelve una cadena recién creada formada a partir de la original eliminando todos los espacios en blanco iniciales.

In [ ]:
# Demostración del método the lstrip()
print("[" + " tau ".lstrip() + "]")

[tau ]


El método con **un parámetro** `lstrip()` hace lo mismo que su versión sin parámetros, pero **elimina todos los caracteres incluidos en el argumento** (una cadena), no solo espacios en blanco:

In [ ]:
print("www.cisco.com".lstrip("w."))

cisco.com


¿Puedes adivinar la salida del fragmento a continuación? Piensa cuidadosamente. Ejecuta el código y verifica tus predicciones.

In [ ]:
print("pythoninstitute.org".lstrip(".org"))

pythoninstitute.org


In [ ]:
# Demostración del método replace()
print("www.netacad.com".replace("netacad.com", "pythoninstitute.org"))
print("This is it!".replace("is", "are"))
print("Apple juice".replace("juice", ""))

www.pythoninstitute.org
Thare are it!
Apple 


La variante del métdodo `replace()` **con tres parámetros** emplea un tercer argumento (un número) para **limitar el número de reemplazos**.

In [ ]:
print("This is it!".replace("is", "are", 1))
print("This is it!".replace("is", "are", 2))

Thare is it!
Thare are it!


Los métodos de uno, dos y tres parámetros denominados `rfind()` hacen casi lo mismo que sus contrapartes (las que carecen del prefijo r), pero **comienzan sus búsquedas desde el final de la cadena**, no el principio (de ahí el prefijo r, de reversa).

In [ ]:
# Demostración del método rfind()
print("tau tau tau".rfind("ta"))
print("tau tau tau".rfind("ta", 9))
print("tau tau tau".rfind("ta", 3, 9))

8
-1
4


In [ ]:
# Demostración del método rstrip()
print("[" + " upsilon ".rstrip() + "]")
print("cisco.com".rstrip(".com"))

[ upsilon]
cis


El método `split()` **divide la cadena y crea una lista de todas las subcadenas detectadas**.

El método **asume que las subcadenas están delimitadas por espacios en blanco** - los espacios no participan en la operación y no se copian en la lista resultante.

N.B.:
- Si la cadena está vacía, la lista resultante también está vacía.
- La operación inversa se puede realizar por el método `join()`.

In [ ]:
# Demostración del método split()
print("phi       chi\npsi".split())

['phi', 'chi', 'psi']


In [ ]:
# Demostración del método startswith()
print("omega".startswith("meg"))
print("omega".startswith("om"))

print()

# Demostración del método strip() 
print("[" + "   aleph   ".strip() + "]")

False
True

[aleph]


In [ ]:
# Demostración del método swapcase()
print("Yo sé que no sé nada.".swapcase())

print()

# Demostración del método title()
print("Yo sé que no sé nada. Parte 1.".title())

print()

# Demostración del método upper()
print("Yo sé que no sé nada. Parte 2.".upper())

yO SÉ QUE NO SÉ NADA.

Yo Sé Que No Sé Nada. Parte 1.

YO SÉ QUE NO SÉ NADA. PARTE 2.


Las **cadenas en Python pueden ser comparadas usando el mismo conjunto de operadores que se emplean con los números**.

Eche un vistazo a estos operadores: también pueden comparar cadenas:

- `==`
- `!=`
- `>`
- `>=`
- `<`
- `<=`

Existe un "pero": los resultados de tales comparaciones a veces pueden ser un poco sorprendentes. No olvides que Python no es consciente (no puede ser de ninguna manera) de problemas lingüísticos sutiles, simplemente **compara valores de puntos de código, caracter por caracter**.

Los resultados que obtienen de una operación de este tipo a veces son sorprendentes. Comencemos con los casos más simples.

Dos cadenas son iguales cuando consisten en los mismos caracteres en el mismo orden. Del mismo modo, dos cadenas no son iguales cuando no consisten en los mismos caracteres en el mismo orden.

Ambas comparaciones dan True (verdadero) como resultado:

In [ ]:
'alfa' == 'alfa'
'alfa' != 'Alfa'

True

La relación final entre cadenas está determinada por **comparar el primer caracter diferente en ambas cadenas** (ten en cuenta los puntos de código ASCII / UNICODE en todo momento).

Cuando se comparan dos cadenas de diferentes longitudes y la más corta es idéntica a la más larga, la **cadena más larga se considera mayor**.

usto como aquí: `'alfa' < 'alfabeto'` es `True` (verdadera).

La comparación de cadenas siempre distingue entre mayúsculas y minúsculas (**las letras mayúsculas se consideran menores en comparación con las minúsculas**). La expresión `'beta' > 'Beta'` es `True` (verdadera).

**Comparar cadenas contra números generalmente es una mala idea**.

Las únicas comparaciones que puede realizar con impunidad son aquellas simbolizadas por los operadores `==` y `!=`. El primero siempre devuelve `False`, mientras que el segundo siempre devuelve `True`.

El uso de cualquiera de los operadores de comparación restantes generará una excepción `TypeError`.

In [ ]:
'10' == 10
'10' != 10
'10' == 1
'10' != 1
#'10' > 10

True

In [ ]:
'10' > 10

TypeError: ignored

Los resultados en este caso son:

In [ ]:
False
True
False
True
TypeError exception

In [ ]:
# Demostración de la función sorted()
firstGreek = ['omega', 'alfa', 'pi', 'gama']
firstGreek2 = sorted(firstGreek)

print(firstGreek)
print(firstGreek2)

print()

# Demostración del método sort()
secondGreek = ['omega', 'alfa', 'pi', 'gama']
print(secondGreek)

secondGreek.sort()
print(secondGreek)

['omega', 'alfa', 'pi', 'gama']
['alfa', 'gama', 'omega', 'pi']

['omega', 'alfa', 'pi', 'gama']
['alfa', 'gama', 'omega', 'pi']


**Cadenas contra números**

Hay dos cuestiones adicionales que deberían discutirse aquí: **cómo convertir un número (un entero o un flotante) en una cadena, y viceversa**. Puede ser necesario realizar tal transformación. Además, es una forma rutinaria de procesar datos de entrada o salida.

La conversión de cadena a número es simple, ya que siempre es posible. Se realiza mediante una función llamada `str()`.

In [ ]:
itg = 13
flt = 1.3
si = str(itg)
sf = str(flt)

print(si + ' ' + sf)

13 1.3


La transformación inversa solo es posible cuando la cadena representa un número válido. Si no se cumple la condición, espera una excepción `ValueError`.

Emplea la función `int()` si deseas obtener un entero, y `float()` si necesitas un valor punto flotante.

In [ ]:
si = '13'
sf = '1.3'
itg = int(si)
flt = float(sf)

print(itg + flt)

14.3
